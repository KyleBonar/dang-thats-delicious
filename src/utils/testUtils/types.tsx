export interface MockLink {
  text: string;
  to: string;
  target: "_blank" | "_self";
}
