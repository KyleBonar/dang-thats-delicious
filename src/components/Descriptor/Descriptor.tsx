import * as React from "react";
import styled from "styled-components";

const StyledH2 = styled.h2`
  font-family: FuturaMedium;
`;

const StyledP = styled.p`
  font-family: AvenirNextReg;
`;

interface DescriptorProps {
  className?: string;
  children: string;
  title: string;
}

const Descriptor: React.SFC<DescriptorProps> = props => {
  return (
    <div className={props.className}>
      <StyledH2>{props.title}</StyledH2>
      <StyledP>{props.children}</StyledP>
    </div>
  );
};

export default Descriptor;
