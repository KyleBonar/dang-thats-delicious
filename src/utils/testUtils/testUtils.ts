import * as React from "react";
import casual from "casual";
import { ReactWrapper } from "enzyme";
import configureStore from "redux-mock-store";

import { MockCard } from "./types";
import { AppState } from "../../redux/configureStore";
import { ISauce, ISaucesState } from "../../redux/sauces/types";
import { FlashMessageProps } from "../../components/FlashMessage/FlashMessage";
import Flatn from "../Flatn/Flatn";
import { IUserState } from "../../redux/users/types";

export const ITERATION_SIZE = 32;
const REACT_REGEX = /react(\d+)?./i;

// seed it so we get consistent results
casual.seed(777);

const relatedSauce = () => ({
  name: casual.name,
  slug: casual.uuid
});

export const fakeSauce = (): ISauce => ({
  _id: casual.random_element([undefined, casual.integer(1, 500)]),
  _addedToStore: casual.random_element([undefined, casual.unix_time]), // Unix time (in seconds) added to redux store
  _full: casual.random_element([undefined, casual.boolean]), // Whether we have full sauce or partial
  _related: casual.random_element([
    undefined,
    new Array(ITERATION_SIZE).fill(null).map(relatedSauce)
  ]), // List of related sauces
  isAdminApproved: casual.random_element([undefined, casual.boolean]),
  name: casual.name,
  ingredients: casual.string,
  author: casual.name,
  created: casual.unix_time, // Unix time (in seconds)
  types: casual.array_of_words(),
  maker: casual.name,
  description: casual.string,
  photo: casual.random_element([undefined, casual.url]),
  shu: casual.random_element([
    undefined,
    casual.integer(1, 500),
    casual.integer(1, 500).toString()
  ]),
  reviews: casual.random_element([
    undefined,
    new Array(ITERATION_SIZE).fill(null).map(x => casual.word)
  ]),
  city: casual.random_element([undefined, casual.city]),
  state: casual.random_element([undefined, casual.state]),
  country: casual.random_element([undefined, casual.country]),
  slug: casual.random_element([undefined, casual.uuid])
});

export const fakeSaucesState = (): ISaucesState => {
  const sauces: ISauce[] = new Array(casual.integer(1, 15))
    .fill(null)
    .map(fakeSauce);
  const { allSlugs, bySlug } = Flatn.saucesArr({ sauces });

  return {
    allSlugs,
    bySlug,
    total: allSlugs.length,
    query: undefined,
    saucesWithNewestReviews: undefined,
    newest: undefined,
    featured: undefined,
    types: casual.array_of_words(),
    orders: ["Newest", "Name", "Times Reviewed", "Avg Rating"]
  };
};

const fakeUsersState = (): IUserState => ({
  self: casual.random_element([
    undefined,
    {
      token: casual.random_element([undefined, casual.uuid]),
      displayName: casual.random_element([undefined, casual.name]),
      avatarURL: casual.random_element([undefined, casual.url]),
      isAdmin: casual.random_element([undefined, casual.boolean])
    }
  ]),
  byDisplayName: undefined,
  allDisplayNames: undefined
});

export const fakeStore = () => {
  const storeConfig = configureStore([]);

  const store: AppState = {
    sauces: fakeSaucesState(),
    users: fakeUsersState(),
    reviews: {
      byReviewID: undefined,
      allReviewIDs: undefined
    }
  };

  return storeConfig(store);
};

export const fakeJSXElement = (): JSX.Element => {
  return casual.random_element([
    React.createElement("div", { key: casual.uuid }, casual.string),
    React.createElement("span", { key: casual.uuid }, casual.string)
  ]);
};

const fakeCard = (): MockCard => ({
  title: casual.string,
  description: casual.string,
  to: casual.url,
  showLink: casual.boolean,
  imageLink: casual.url,
  className: casual.string,
  anchorText: casual.string
});

const wait = (amount = 0) =>
  new Promise(resolve => setTimeout(resolve, amount));

function simulateInputChange(
  wrapper: ReactWrapper,
  name: string,
  value: string
) {
  wrapper
    // .find(`input[name='${name}']`)
    .simulate("change", { target: { name, value } });
}

const isClassComponent = (component): boolean => {
  return (
    typeof component === "function" &&
    component.prototype &&
    !!component.prototype.isReactComponent
  );
};

// Ensure compatability with transformed code
const isFunctionComponent = (component): boolean => {
  return (
    typeof component === "function" &&
    String(component).includes("return") &&
    !!String(component).match(REACT_REGEX) &&
    String(component).includes(".createElement")
  );
};

const isElement = (typeElement): boolean => {
  return React.isValidElement(typeElement);
};

const isComponent = (component): boolean => {
  return isClassComponent(component) || isFunctionComponent(component);
};

const isDOMTypeElement = (typeElement): boolean => {
  return isElement(typeElement) && typeof typeElement.type === "string";
};

const generateValidPassword = (MIN_PASSWORD_LENGTH: number = 8) => {
  // generate length between minimum length and 3x the minimum length
  const _randomLength = casual.integer(
    MIN_PASSWORD_LENGTH,
    3 * MIN_PASSWORD_LENGTH
  );

  // turn generated length into random letters
  return new Array(_randomLength).fill(null).map(casual._letter).join("");
};

const generateInValidPassword = (MIN_PASSWORD_LENGTH: number = 8) => {
  // generate length between minimum length and 3x the minimum length
  const _randomLength = casual.integer(0, MIN_PASSWORD_LENGTH);

  // turn generated length into random letters
  return new Array(_randomLength).fill(null).map(casual._letter).join("");
};

export {
  casual,
  simulateInputChange,
  isDOMTypeElement,
  isComponent,
  wait,
  generateValidPassword
};
