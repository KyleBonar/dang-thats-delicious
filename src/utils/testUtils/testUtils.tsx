import * as React from "react";
import * as enzyme from "enzyme";
import casual from "casual";
import { ReactWrapper } from "enzyme";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import { AppState } from "../../redux/configureStore";
import { ISauce, ISaucesState } from "../../redux/sauces/types";
import Flatn from "../Flatn/Flatn";
import { IUserState } from "../../redux/users/types";
import { IErrReturn } from "../Err/Err";
import { act } from "react-dom/test-utils";
import { Provider } from "react-redux";

export const ITERATION_SIZE = 32;
const REACT_REGEX = /react(\d+)?./i;

// seed it so we get consistent results
casual.seed(777);

const relatedSauce = () => ({
  name: casual.name,
  slug: casual.uuid
});

export const fakeSauce = (): ISauce => ({
  _id: casual.random_element([undefined, casual.integer(1, 500)]),
  _addedToStore: casual.random_element([undefined, casual.unix_time]), // Unix time (in seconds) added to redux store
  _full: casual.random_element([undefined, casual.boolean]), // Whether we have full sauce or partial
  _related: casual.random_element([
    undefined,
    new Array(ITERATION_SIZE).fill(null).map(relatedSauce)
  ]), // List of related sauces
  isAdminApproved: casual.random_element([undefined, casual.boolean]),
  name: casual.name,
  ingredients: casual.random_element([undefined, casual.string]),
  author: casual.name,
  created: casual.unix_time, // Unix time (in seconds)
  types: casual.random_element([undefined, casual.array_of_words()]),
  maker: casual.name,
  description: casual.string,
  photo: casual.random_element([undefined, casual.url]),
  shu: casual.random_element([
    undefined,
    casual.integer(1, 500),
    casual.integer(1, 500).toString()
  ]),
  reviews: casual.random_element([
    undefined,
    new Array(ITERATION_SIZE).fill(null).map(x => casual.word)
  ]),
  city: casual.random_element([undefined, casual.city]),
  state: casual.random_element([undefined, casual.state]),
  country: casual.random_element([undefined, casual.country]),
  slug: casual.random_element([undefined, casual.uuid])
});

export const generateFakeSauces = (): ISauce[] =>
  new Array(casual.integer(1, 15)).fill(null).map(fakeSauce);

export const fakeSaucesState = (): ISaucesState => {
  const sauces: ISauce[] = generateFakeSauces();
  const { allSlugs, bySlug } = Flatn.saucesArr({ sauces });

  const featured: undefined | string[] = casual.random_element([
    undefined,
    allSlugs.length > 0
      ? allSlugs.filter(slug => {
          if (casual.random < 0.4) return slug;
          return;
        })
      : undefined
  ]);

  return {
    allSlugs,
    bySlug,
    total: allSlugs.length,
    query: undefined,
    saucesWithNewestReviews: undefined,
    newest: undefined,
    featured,
    types: casual.array_of_words(),
    orders: ["Newest", "Name", "Times Reviewed", "Avg Rating"]
  };
};

const fakeUsersState = (): IUserState => ({
  self: casual.random_element([
    undefined,
    {
      token: casual.random_element([undefined, casual.uuid]),
      displayName: casual.random_element([undefined, casual.name]),
      avatarURL: casual.random_element([undefined, casual.url]),
      isAdmin: casual.random_element([undefined, casual.boolean])
    }
  ]),
  byDisplayName: undefined,
  allDisplayNames: undefined
});

export const fakeStore = () => {
  const middlewares = [thunk];
  const storeConfig = configureStore(middlewares);

  const store: AppState = {
    sauces: fakeSaucesState(),
    users: fakeUsersState(),
    reviews: {
      byReviewID: undefined,
      allReviewIDs: undefined
    }
  };

  return storeConfig(store);
};

export const fakeJSXElement = (): JSX.Element => {
  return casual.random_element([
    React.createElement("div", { key: casual.uuid }, casual.string),
    React.createElement("span", { key: casual.uuid }, casual.string)
  ]);
};

const wait = (amount = 0) =>
  new Promise(resolve => setTimeout(resolve, amount));

function simulateInputChange(
  wrapper: ReactWrapper,
  name: string,
  value: string
) {
  wrapper
    // .find(`input[name='${name}']`)
    .simulate("change", { target: { name, value } });
}

const isClassComponent = (component): boolean => {
  return (
    typeof component === "function" &&
    component.prototype &&
    !!component.prototype.isReactComponent
  );
};

// Ensure compatability with transformed code
const isFunctionComponent = (component): boolean => {
  return (
    typeof component === "function" &&
    String(component).includes("return") &&
    !!String(component).match(REACT_REGEX) &&
    String(component).includes(".createElement")
  );
};

const isElement = (typeElement): boolean => {
  return React.isValidElement(typeElement);
};

const isComponent = (component): boolean => {
  return isClassComponent(component) || isFunctionComponent(component);
};

const isDOMTypeElement = (typeElement): boolean => {
  return isElement(typeElement) && typeof typeElement.type === "string";
};

const generateValidPassword = (MIN_PASSWORD_LENGTH: number = 8): string => {
  // generate length between minimum length and 3x the minimum length
  const _randomLength = casual.integer(
    MIN_PASSWORD_LENGTH,
    3 * MIN_PASSWORD_LENGTH
  );

  // turn generated length into random letters
  return new Array(_randomLength).fill(null).map(casual._letter).join("");
};

const generateInValidPassword = (MIN_PASSWORD_LENGTH: number = 8): string => {
  // generate length between minimum length and 3x the minimum length
  const _randomLength = casual.integer(1, MIN_PASSWORD_LENGTH - 1);

  // turn generated length into random letters
  return new Array(_randomLength).fill(null).map(casual._letter).join("");
};

// Function designed to resemble how a network error would look like.
// This will help the catch(err) only have to handle a single format
const generateErr = (): IErrReturn => {
  return {
    response: {
      data: {
        status: 400,
        msg: casual.string,
        isGood: false
      }
    }
  };
};

// lets us mount up a simple custom hook
const mountReactHook = hook => {
  const Component = ({ children }) => children(hook());
  const componentHook = {};
  let componentMount;

  act(() => {
    componentMount = enzyme.shallow(
      <Component>
        {hookValues => {
          Object.assign(componentHook, hookValues);
          return null;
        }}
      </Component>
    );
  });
  return { componentMount, componentHook };
};

export interface ImountReactHookWithReduxStore {
  componentMount: any;
  componentHook: any;
}
const mountReactHookWithReduxStore = (
  hook,
  store
): ImountReactHookWithReduxStore => {
  const Component = ({ children }) => children(hook());
  const componentHook = {};
  let componentMount;

  act(() => {
    componentMount = enzyme.mount(
      <Provider store={store}>
        <Component>
          {hookValues => {
            Object.assign(componentHook, hookValues);
            return null;
          }}
        </Component>
      </Provider>
    );
  });
  return { componentMount, componentHook };
};

export {
  casual,
  simulateInputChange,
  isDOMTypeElement,
  isComponent,
  wait,
  generateValidPassword,
  generateInValidPassword,
  generateErr,
  mountReactHook,
  mountReactHookWithReduxStore
};
