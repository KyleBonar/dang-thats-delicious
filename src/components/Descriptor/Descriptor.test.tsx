import * as React from "react";
import * as enzyme from "enzyme";

import { casual, ITERATION_SIZE } from "../../utils/testUtils/testUtils";
import Descriptor, { DescriptorProps } from "./Descriptor";

const fakeDescriptor = (): DescriptorProps => ({
  className: casual.random_element([undefined, casual.string]),
  children: casual.string,
  title: casual.string
});

const mockDescriptors = new Array(ITERATION_SIZE)
  .fill(null)
  .map(fakeDescriptor);

describe("<Descriptor>", () => {
  it("renders", () => {
    mockDescriptors.forEach(mockDescriptor => {
      const wrapper = enzyme.shallow(<Descriptor {...mockDescriptor} />);
      expect(wrapper).toBeTruthy();
    });
  });

  it("matches snapshot", () => {
    mockDescriptors.forEach(mockDescriptor => {
      const wrapper = enzyme.shallow(<Descriptor {...mockDescriptor} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  it("renders expected title", () => {
    mockDescriptors.forEach(mockDescriptor => {
      const wrapper = enzyme.shallow(<Descriptor {...mockDescriptor} />);

      expect(wrapper.find("h2").text()).toEqual(mockDescriptor.title);
    });
  });

  it("renders expected description", () => {
    mockDescriptors.forEach(mockDescriptor => {
      const wrapper = enzyme.shallow(<Descriptor {...mockDescriptor} />);

      expect(wrapper.find("p").text()).toEqual(mockDescriptor.children);
    });
  });

  it("adds className to parent", () => {
    mockDescriptors.forEach(mockDescriptor => {
      if (!mockDescriptor.className) return;

      const wrapper = enzyme.shallow(<Descriptor {...mockDescriptor} />);

      expect(wrapper.find("div").prop("className")).toContain(
        mockDescriptor.className
      );
    });
  });
});
