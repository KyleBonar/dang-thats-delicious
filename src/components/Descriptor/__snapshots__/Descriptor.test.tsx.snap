// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Descriptor> matches snapshot 1`] = `
<div>
  <h2>
    aut enim iure facilis ut explicabo iure
  </h2>
  <p>
    fugiat error molestiae laborum est possimus assumenda
  </p>
</div>
`;

exports[`<Descriptor> matches snapshot 2`] = `
<div
  className="vitae occaecati officia iure nisi ut maiores"
>
  <h2>
    quia autem et modi tenetur amet modi
  </h2>
  <p>
    fuga reprehenderit soluta non omnis consequatur enim
  </p>
</div>
`;

exports[`<Descriptor> matches snapshot 3`] = `
<div>
  <h2>
    ipsa tempora aut sint aperiam occaecati illum
  </h2>
  <p>
    dolorum ab aliquam id sit delectus vel
  </p>
</div>
`;

exports[`<Descriptor> matches snapshot 4`] = `
<div
  className="animi vel provident dicta perferendis dolorum cupiditate"
>
  <h2>
    omnis laborum exercitationem est sapiente laudantium molestias
  </h2>
  <p>
    saepe ut assumenda sint quia sed exercitationem
  </p>
</div>
`;

exports[`<Descriptor> matches snapshot 5`] = `
<div
  className="assumenda quasi adipisci mollitia iste itaque omnis"
>
  <h2>
    corporis commodi eaque nisi cum ducimus nemo
  </h2>
  <p>
    rem similique temporibus aliquam sed sapiente soluta
  </p>
</div>
`;

exports[`<Descriptor> matches snapshot 6`] = `
<div>
  <h2>
    autem sit aspernatur aspernatur quam eligendi est
  </h2>
  <p>
    ipsam laboriosam non itaque voluptas ea ducimus
  </p>
</div>
`;

exports[`<Descriptor> matches snapshot 7`] = `
<div>
  <h2>
    sit suscipit omnis necessitatibus maiores ullam quae
  </h2>
  <p>
    ex expedita est doloremque explicabo eaque necessitatibus
  </p>
</div>
`;

exports[`<Descriptor> matches snapshot 8`] = `
<div
  className="corporis qui aut nobis quo consequatur harum"
>
  <h2>
    nulla aut sunt est rerum at sint
  </h2>
  <p>
    enim et doloremque voluptate in sunt occaecati
  </p>
</div>
`;
